File: backend/members/views_frontend/address_lookup.py
import requests
from django.http import JsonResponse


def postcode_lookup(request):
    postcode = request.GET.get("postcode", "").strip()

    if not postcode:
        return JsonResponse({"error": "Postcode required"}, status=400)

    url = f"https://api.postcodes.io/postcodes/{postcode}"

    try:
        response = requests.get(url, timeout=5)
        data = response.json()

        if data["status"] != 200:
            return JsonResponse({"error": "Invalid postcode"}, status=400)

        result = data["result"]

        return JsonResponse({
            "line_1": result.get("admin_ward"),
            "town": result.get("admin_district"),
            "county": result.get("region"),
            "country": result.get("country"),
        })

    except Exception:
        return JsonResponse({"error": "Lookup failed"}, status=500)

File: backend/static/js/address_lookup.js
function lookupPostcode() {
    const postcode = document.getElementById("postcode").value;

    if (!postcode) {
        alert("Enter a postcode");
        return;
    }

    fetch(`/api/postcode-lookup/?postcode=${postcode}`)
        .then(res => res.json())
        .then(data => {
            if (data.success) {
                document.getElementById("town").value = data.town || "";
                document.getElementById("county").value = data.county || "";
            } else {
                alert("Postcode not found");
            }
        })
        .catch(() => {
            alert("Error looking up postcode");
        });
}